ledger/accountdb_test.go

```diff
@@ -785,3 +789,81 @@ func TestAccountsDbQueriesCreateClose(t *testing.T) {
 	qs.close()
 	require.Nil(t, qs.listCreatablesStmt)
 }
+
+func benchmarkWriteCatchpointStagingBalancesSub(b *testing.B) {
+	genesisInitState, _ := testGenerateInitState(b, protocol.ConsensusCurrentVersion)
+	const inMem = false
+	log := logging.TestingLog(b)
+	cfg := config.GetDefaultLocal()
+	cfg.Archival = false
+	log.SetLevel(logging.Warn)
+	dbBaseFileName := strings.Replace(b.Name(), "/", "_", -1)
+	l, err := OpenLedger(log, dbBaseFileName, inMem, genesisInitState, cfg)
+	require.NoError(b, err, "could not open ledger")
+	defer func() {
+		l.Close()
+		os.Remove(dbBaseFileName + ".block.sqlite")
+		os.Remove(dbBaseFileName + ".tracker.sqlite")
+	}()
+	catchpointAccessor := MakeCatchpointCatchupAccessor(l, log)
+	catchpointAccessor.ResetStagingBalances(context.Background(), true)
+	accountsCount := uint64(b.N)
+	b.ResetTimer()
+	accounts := uint64(0)
+	var last64KStart time.Time
+	last64KSize := uint64(0)
+	last64KSubTime := time.Duration(0)
+	accountsWritingStarted := time.Now()
+	for accounts < accountsCount {
+		b.StopTimer()
+		balancesLoopStart := time.Now()
+		// generate a chunk;
+		chunkSize := accountsCount - accounts
+		if chunkSize > BalancesPerCatchpointFileChunk {
+			chunkSize = BalancesPerCatchpointFileChunk
+		}
+		last64KSize += chunkSize
+		if accounts >= accountsCount-64*1024 && last64KStart.IsZero() {
+			last64KStart = time.Now()
+			last64KSize = chunkSize
+			last64KSubTime = time.Duration(0)
+		}
+		var balances catchpointFileBalancesChunk
+		balances.Balances = make([]encodedBalanceRecord, chunkSize)
+		for i := uint64(0); i < chunkSize; i++ {
+			var randomAccount encodedBalanceRecord
+			accountData := basics.AccountData{RewardsBase: accounts + i}
+			accountData.MicroAlgos.Raw = crypto.RandUint63()
+			randomAccount.AccountData = protocol.Encode(&accountData)
+			crypto.RandBytes(randomAccount.Address[:])
+			balances.Balances[i] = randomAccount
+		}
+		last64KSubTime += time.Now().Sub(balancesLoopStart)
+		b.StartTimer()
+		err = l.trackerDBs.wdb.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+			err = writeCatchpointStagingBalances(ctx, tx, balances.Balances)
+			return
+		})
+
+		require.NoError(b, err)
+		accounts += chunkSize
+	}
+	if !last64KStart.IsZero() {
+		last64KDuration := time.Now().Sub(last64KStart) - last64KSubTime
+		fmt.Printf("%-74s%-7d (last 64k) %-6d ns/account       %d accounts/sec\n", b.Name(), last64KSize, (last64KDuration / time.Duration(last64KSize)).Nanoseconds(), int(float64(last64KSize)/float64(last64KDuration.Seconds())))
+	}
+	stats, err := l.trackerDBs.wdb.Vacuum(context.Background())
+	require.NoError(b, err)
+	fmt.Printf("%-74sdb fragmentation   %.1f%%\n", b.Name(), float32(stats.PagesBefore-stats.PagesAfter)*100/float32(stats.PagesBefore))
+	b.ReportMetric(float64(b.N)/float64(time.Now().Sub(accountsWritingStarted).Seconds()), "accounts/sec")
```

Comment:
The measurement here includes the account creation time, but the op/ns reported on the same line is just the database operation time. Also, it is not consistent with last64k calculation.
Hence, ns/op is faster than accounts/sec.

e.g. 
BenchmarkWriteCatchpointStagingBalances/Restore-102400                    65536   (last 64k) 66068  ns/account       15135 accounts/sec
BenchmarkWriteCatchpointStagingBalances/Restore-102400                    db fragmentation   3.7%
BenchmarkWriteCatchpointStagingBalances/Restore-102400-8         	  102400	     **58041** ns/op	     17284 accounts/sec

58041 should be 1/17284*1e9=57857

Reply:
I've updated the calculation per your notes. I wasn't able to see any notable change.

Reply:
It won't make much difference, because the extra account creation time is so tiny compared to the database operations. However, when database operations are optimized, the difference may become more significant. 
Thanks for making the change.